content/style-injector.js

```diff
@@ -1,184 +1,335 @@
+    if (el.parentNode !== creationDoc.documentElement) {
+      return true;
+    }
+    let i = 0;
+    while (el) {
+      if (i < list.length && el === list[i].el) {
+        i++;
+      } else if (ORDERED_TAGS.has(el.localName)) {
+        return true;
+      }
+      el = el.nextElementSibling;
+    }
+    // some styles are not injected to the document
+    if (i < list.length) {
+      _sort();
+      return true;
+    }
+  }
+
+  function _styleMapToArray(styleMap) {
+    return Object.values(styleMap).map(s => ({
+      id: s.id,
+      code: s.code.join(''),
+    }));
+  }
+
+  function _update({id, code}) {
+    const style = table.get(id);
+    if (style.code === code) return;
+    style.code = code;
+    // workaround for Chrome devtools bug fixed in v65
+    if (isChromePre65) {
+      const oldEl = style.el;
+      style.el = _createStyle(id, code);
+      oldEl.parentNode.insertBefore(style.el, oldEl.nextSibling);
+      oldEl.remove();
+    } else {
+      style.el.textContent = code;
     }
+    // https://github.com/openstyles/stylus/issues/693
+    style.el.disabled = !isEnabled;
   }
 
-  function emitUpdate() {
-    if (onUpdate) {
-      onUpdate();
+  function RewriteObserver(onChange) {
+    // detect documentElement being rewritten from inside the script
+    let root;
+    let observing = false;
+    let timer;
+    const observer = new MutationObserver(_check);
+    return {start, stop};
+
+    function start() {
+      if (observing) return;
+      // detect dynamic iframes rewritten after creation by the embedder i.e. externally
+      root = document.documentElement;
+      timer = setTimeout(_check);
+      observer.observe(document, {childList: true});
+      observing = true;
+    }
+
+    function stop() {
+      if (!observing) return;
+      clearTimeout(timer);
+      observer.disconnect();
+      observing = false;
+    }
+
+    function _check() {
+      if (root !== document.documentElement) {
+        root = document.documentElement;
+        onChange();
+      }
     }
   }
 
-  function replaceAll(styles) {
-    const added = new Set(styles.map(s => s.id));
-    const removed = [];
-    for (const style of list) {
-      if (!added.has(style.id)) {
-        removed.push(style.id);
+  function DocRootObserver(onChange) {
+    let digest = 0;
+    let lastCalledTime = NaN;
+    let observing = false;
+    const observer = new MutationObserver(() => {
+      if (digest) {
+        if (performance.now() - lastCalledTime > 1000) {
+          digest = 0;
+        } else if (digest > 5) {
+          throw new Error('The page keeps generating mutations. Skip the event.');
+        }
       }
+      if (onChange()) {
+        digest++;
+        lastCalledTime = performance.now();
+      }
+    });
+    return {evade, start, stop};
+
+    function evade(fn) {
+      const restore = observing && start;
+      stop();
+      return new Promise(resolve => _run(fn, resolve, _waitForRoot))
+        .then(restore);
```

Comment:
Should we avoid passing non-callable to `Promise.then`? It might confuse some developers.

Reply:
Those developers would have a chance to read the spec or at least [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then) like I did :-) and get to know that it's perfectly fine: "If one or both arguments are omitted or are provided non-functions, then then will be missing the handler(s), but will not generate any errors."